src/Aspire.Hosting/ExecutableResourceBuilderExtensions.cs

```diff
@@ -69,30 +79,64 @@ public static IResourceBuilder<ExecutableResource> AddExecutable(this IDistribut
+            {
+                c.WithBuildArg(arg.Name, arg.Value);
+            }
+        });
     }
 
-    private static async Task WriteExecutableAsDockerfileResourceAsync(ManifestPublishingContext context, ExecutableResource executable, IEnumerable<DockerBuildArg>? buildArgs = null)
+    /// <summary>
+    /// Adds support for containerizating this <see cref="ExecutableResource"/> during deployment.
+    /// The resulting container image is built, and when the optional <paramref name="configure"/> action is provided,
+    /// it is used to configure the container resource.
+    /// </summary>
+    /// <typeparam name="T">Type of executable resource</typeparam>
+    /// <param name="builder">Resource builder</param>
+    /// <param name="configure">Optional action to configure the container resource</param>
+    /// <returns>A reference to the <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<T> PublishAsDockerFile<T>(this IResourceBuilder<T> builder, Action<IResourceBuilder<ContainerResource>>? configure)
+        where T : ExecutableResource
     {
-        context.Writer.WriteString("type", "dockerfile.v0");
+        if (!builder.ApplicationBuilder.ExecutionContext.IsPublishMode)
+        {
+            return builder;
+        }
 
-        var appHostRelativePathToDockerfile = Path.Combine(executable.WorkingDirectory, "Dockerfile");
-        var manifestFileRelativePathToDockerfile = context.GetManifestRelativePath(appHostRelativePathToDockerfile);
-        context.Writer.WriteString("path", manifestFileRelativePathToDockerfile);
+        // The implementation here is less than ideal, but we don't have a clean way of building resource types
+        // that change their behavior based on the context. In this case, we want to change the behavior of the
+        // resource from an ExecutableResource to a ContainerResource. We do this by removing the ExecutableResource
+        // from the application model and adding a new ContainerResource in its place in publish mode.
 
-        var manifestFileRelativePathToContextDirectory = context.GetManifestRelativePath(executable.WorkingDirectory);
-        context.Writer.WriteString("context", manifestFileRelativePathToContextDirectory);
+        // There are still dangling references to the original ExecutableResource in the application model, but
+        // in publish mode, it won't be used. This is a limitation of the current design.
+        builder.ApplicationBuilder.Resources.Remove(builder.Resource);
 
-        if (buildArgs is not null)
-        {
-            context.WriteDockerBuildArgs(buildArgs);
-        }
+        var container = new ExecutableContainerResource(builder.Resource);
+        var cb = builder.ApplicationBuilder.AddResource(container);
+        cb.WithImage(builder.Resource.Name);
+        cb.WithDockerfile(contextPath: builder.Resource.WorkingDirectory);
+        // Clear the runtime args
+        cb.WithArgs(c => c.Args.Clear());
```

Comment:
Why? Can we say why in the comment?

Reply:
Yea I should add a comment here. The args that apply to the executable *could* in theory apply to the container but for the ones we ship it does not just work.